### PR TITLE
[BPF] avoid calling bpftool and tc to check if things changed

### PR DESCRIPTION
**Pull request by tomastigera**
## Description
   [BPF] use libbpf to detach tc programs rather than tc itself

    [BPF] avoid calling bpftool and tc to check if things changed
    
    Collect information on all attached programs at the top of
    CompleteDeferredWork and remember that for the duration of it. That is a
    single call to bpftool if there is any change instead of calling and
    parsing it for each modified device.
    
    We also called the bpftool and tc tool multiple times when things
    changed, especially when felix restarted.

    [BPF] xpd AlreadyAttached() calls ProgramID 2x
    
    There is no need for that, if we cannot get program id we know that
    our stuff it not attached and we dont care of something else is
    attached. We need to attach ourselves.

    felix/Makefile: BPF ut test may need roo and cgo

    [BPF] bpf_ep_mgr jumpMapFDs are an array not a map

    [BPF] convert bpf.Hook from string to int
    
    Less need for maps as indexing becomes easy.

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
TBD
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

**Review by fasaxc — CHANGES_REQUESTED** [review]

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 1]
```diff
@@ -260,6 +260,8 @@ type bpfEndpointManager struct {
 	natOutIdx int
 
 	arpMap bpf.Map
+
+	attachedProgs map[bpf.HookDevKey]int
```
[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:264]: yup, that is where it is being cleared

File: felix/dataplane/linux/bpf_ep_mgr.go [hunk 2]
```diff
@@ -1063,19 +1100,27 @@ func (m *bpfEndpointManager) doApplyPolicy(ifaceName string, isReady *bool) erro
...
 			log.WithField("iface", ifaceName).Debug("Interface is down/gone, closing jump maps.")
-			for _, fd := range iface.dpState.jumpMapFDs {
+			for i, fd := range iface.dpState.jumpMapFDs {
+				if fd == 0 {
+					continue
+				}
 				if err := fd.Close(); err != nil {
 					log.WithError(err).Error("Failed to close jump map.")
 				}
+				iface.dpState.jumpMapFDs[i] = 0
```
[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/dataplane/linux/bpf_ep_mgr.go:1122]: shouldn't the deep equal match that the array is zeroed?

File: felix/bpf/attach.go [hunk 3]
```diff
@@ -271,3 +273,60 @@ func ListCalicoAttached() (map[string]EPAttachInfo, error) {
...
+type HookDevKey struct {
+	Dev  string
+	Hook Hook
+}
+
+func ListHooksIDs() map[HookDevKey]int {
+	// Find all the programs that are attached to interfaces.
+	out, err := exec.Command("bpftool", "-j", "net").Output()
+	if err != nil {
+		log.WithError(err).Panic("Failed to list attached bpf programs")
```
[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/bpf/attach.go:286]: sure, I think I went with copy paste, but error it fine

File: felix/bpf/libbpf/libbpf.go [hunk 4]
```diff
@@ -121,8 +121,13 @@ func (m *Map) NextMap() (*Map, error) {
 	return &Map{bpfMap: bpfMap, bpfObj: m.bpfObj}, nil
 }
 
-func (o *Obj) AttachClassifier(secName, ifName, hook string) (int, error) {
-	isIngress := 0
+func DetachClassifier(ifindex, handle, pref int, ingress bool) error {
```
[the remark this answers is not in the document]
Comment (tomastigera) [reply felix/bpf/libbpf/libbpf.go:130]: but attach is method of object and Query/Detach are functions unrelated to obj, so I placed it in front, but next to it :man_shrugging:

**Comment by tomastigera** [comment]
Superseded by https://github.com/projectcalico/calico/pull/7219

**Pull request closed without merging**